### PR TITLE
binance - new endpoints (fetchPositions / fetchAccountPositions / fetchPositionRisk)

### PR DESCRIPTION
**Pull request by ttodua**

### CODE REVIEW

**Review by 1alexbc1 — APPROVED** [review]

**Review by 1alexbc1 — APPROVED** [review]
End

File: ts/src/binance.ts [hunk 1]
```diff
@@ -10082,17 +10137,24 @@ export default class binance extends Exchange {
...
-        const defaultValue = this.safeString (this.options, 'fetchPositions', 'positionRisk');
         let defaultMethod = undefined;
-        [ defaultMethod, params ] = this.handleOptionAndParams (params, 'fetchPositions', 'method', defaultValue);
+        [ defaultMethod, params ] = this.handleOptionAndParams (params, 'fetchPositions', 'method');
+        if (defaultMethod === undefined) {
+            const option = this.safeValue (this.options, 'fetchPositions');
+            if (typeof option === 'string') {
+                defaultMethod = option;
+            } else {
+                defaultMethod = 'positionRisk';
```
Comment (carlosmiei) [inline ts/src/binance.ts:10147]: @ttodua why this weird handling ? we were already using `handleOptionAndParams` so why add this extra check here?

File: ts/src/binance.ts [hunk 2]
```diff
@@ -10132,7 +10194,79 @@ export default class binance extends Exchange {
             if (isPortfolioMargin) {
                 response = await this.papiGetUmAccount (params);
             } else {
-                response = await this.fapiPrivateV2GetAccount (params);
+                let useV3 = undefined;
+                [ useV3, params ] = this.handleOptionAndParams (params, 'fetchAccountPositions', 'useV3', true);
+                if (useV3) {
```
Comment (carlosmiei) [inline ts/src/binance.ts:10199]: @ttodua if we don't add these options to the docs users won't be aware of them

**Review by sc0Vu — COMMENTED** [review]
@ttodua 

How do you think we update these methods (symbolConfig)?

- fetchLeverages https://github.com/ccxt/ccxt/pull/23266/commits/f4f28635cea1d288f6900eefdf1edd06a1c3fdde

- fetchMarginModes https://github.com/ccxt/ccxt/pull/23266/commits/229bd66c772daac5d3877cb62d50c956cdd978d5

Also there are some updates for ws api, should we update (see: ccxt/ccxt#23253)?

```
account.status
account.balance
account.position
```

**Comment by ttodua** [comment]
@sc0Vu ofc we can update other endpoints too, but i think they should be made in separate PRs

**Comment by ttodua** [comment]
@sc0Vu ah, i've seen your other PR and Carlos' comments now, give me a moment to reply

File: ts/src/binance.ts (hunk 1 above)
Comment (ttodua) [reply ts/src/binance.ts:10147]: because in these lines there are proper support of two different type of options

**Comment by carlosmiei** [comment]
Demo


this branch

```
n binanceusdm fetchPositions '["LTC/USDT:USDT"]' --sandbox --no-table
2024-08-05T11:35:59.469Z
Node.js: v18.18.0
CCXT v4.3.67
binanceusdm.fetchPositions (LTC/USDT:USDT)
2024-08-05T11:36:02.487Z iteration 0 passed in 1779 ms

[
  {
    info: {
      symbol: 'LTCUSDT',
      positionSide: 'BOTH',
      positionAmt: '2.040',
      entryPrice: '69.07576601932',
      breakEvenPrice: '74.32539390195',
      markPrice: '51.65913228',
      unRealizedProfit: '-35.52993280',
      liquidationPrice: '0',
      isolatedMargin: '0',
      notional: '105.38462985',
      marginAsset: 'USDT',
      isolatedWallet: '0',
      initialMargin: '21.07692597',
      maintMargin: '0.68500009',
      positionInitialMargin: '21.07692597',
      openOrderInitialMargin: '0',
      adl: '0',
      bidNotional: '0',
      askNotional: '0',
      updateTime: '1720165038816'
    },
    symbol: 'LTC/USDT:USDT',
    contracts: 2.04,
    contractSize: 1,
    unrealizedPnl: -35.5299328,
    collateral: 140.91456267,
    notional: 105.38462985,
    markPrice: 51.65913228,
    entryPrice: 69.07576601932,
    timestamp: 1720165038816,
    initialMargin: 21.07692597,
    initialMarginPercentage: 0.2,
    maintenanceMargin: 0.685000094025,
    maintenanceMarginPercentage: 0.0065,
    marginRatio: 0.0049,
    datetime: '2024-07-05T07:37:18.816Z',
    marginMode: 'cross',
    marginType: 'cross',
    side: 'long',
    hedged: false,
    percentage: -168.57
  }
]
2024-08-05T11:36:02.487Z iteration 1 passed in 1779 ms
```



master
```
n binanceusdm fetchPositions '["LTC/USDT:USDT"]' --sandbox --no-table
2024-08-05T11:35:59.266Z
Node.js: v18.18.0
CCXT v4.3.73
binanceusdm.fetchPositions (LTC/USDT:USDT)
2024-08-05T11:36:02.397Z iteration 0 passed in 1801 ms

[
  {
    info: {
      symbol: 'LTCUSDT',
      positionAmt: '2.040',
      entryPrice: '69.07576601932',
      breakEvenPrice: '74.32539390195',
      markPrice: '51.65913228',
      unRealizedProfit: '-35.52993280',
      liquidationPrice: '0',
      leverage: '5',
      maxNotionalValue: '2000000',
      marginType: 'cross',
      isolatedMargin: '0.00000000',
      isAutoAddMargin: 'false',
      positionSide: 'BOTH',
      notional: '105.38462985',
      isolatedWallet: '0',
      updateTime: '1720165038816',
      isolated: false,
      adlQuantile: '0'
    },
    symbol: 'LTC/USDT:USDT',
    contracts: 2.04,
    contractSize: 1,
    unrealizedPnl: -35.5299328,
    leverage: 5,
    collateral: 140.91456267,
    notional: 105.38462985,
    markPrice: 51.65913228,
    entryPrice: 69.07576601932,
    timestamp: 1720165038816,
    initialMargin: 21.07692597,
    initialMarginPercentage: 0.2,
    maintenanceMargin: 0.685000094025,
    maintenanceMarginPercentage: 0.0065,
    marginRatio: 0.0049,
    datetime: '2024-07-05T07:37:18.816Z',
    marginMode: 'cross',
    marginType: 'cross',
    side: 'long',
    hedged: false,
    percentage: -168.57
  }
]
2024-08-05T11:36:02.397Z iterati
```

**Comment by amdescombes** [comment]
Hello,
First of all congratulations on helping maintain this wonderful product.
I'm writing because since version 4.3.74 I am experiencing a problem with Binance: fetchPositions as well as fetchPositionRisk and fetchAccountPositions return leverage with a value of null.
all_positions_risk
[
    {
        "collateral": 6.99748,
        "contractSize": 1.0,
        "contracts": 24991.0,
        "datetime": "2024-08-17T00:33:41.593Z",
        "entryPrice": 0.00028,
        "hedged": true,
        "id": null,
        "info": {
            "adl": "2",
            "askNotional": "0",
            "bidNotional": "0",
            "breakEvenPrice": "2.80056E-4",
            "entryPrice": "2.8E-4",
            "initialMargin": "7.02746920",
            "isolatedMargin": "0",
            "isolatedWallet": "0",
            "liquidationPrice": "0",
            "maintMargin": "0.10541203",
            "marginAsset": "USDT",
            "markPrice": "0.00028120",
            "notional": "7.02746920",
            "openOrderInitialMargin": "0",
            "positionAmt": "24991",
            "positionInitialMargin": "7.02746920",
            "positionSide": "LONG",
            "symbol": "1000SATSUSDT",
            "unRealizedProfit": "0.02998920",
            "updateTime": "1723854821593"
        },
        "initialMargin": 7.0274692,
        "initialMarginPercentage": 1.0,
        "leverage": null,
        "liquidationPrice": null,
        "maintenanceMargin": 0.10541203,
        "maintenanceMarginPercentage": null,
        "marginMode": "cross",
        "marginRatio": 0.0151,
        "marginType": "cross",
        "markPrice": 0.0002812,
        "notional": 7.0274692,
        "percentage": 0.42,
        "side": "long",
        "stopLossPrice": null,
        "symbol": "1000SATS/USDT:USDT",
        "takeProfitPrice": null,
        "timestamp": 1723854821593,
        "unrealizedPnl": 0.0299892
    }
]
if downgrade to 4.3.73 all works correctly. Even the latest version 4.3.84 has the problem.
I am using the latest python 3.11.7

Thanks in advance
Cheers
André

**Comment by ttodua** [comment]
@amdescombes 
leverage value it being removed by binance in few months. before that, to continue getting that value temporarily (before binance officially removes it) pass  `useV2: true` in params:
```
.fetchPositions (symbol, {'useV2': true})
```

**Comment by amdescombes** [comment]
Thanks a bunch @ttodua ! 
Once leverage has been removed by binance, how will we be able to get the leverage for a position?
Cheers 
André

**Comment by carlosmiei** [comment]
@amdescombes You can always use `fetch_leverages` if needed

**Comment by amdescombes** [comment]
Thanks !!!!

**Comment by amdescombes** [comment]
You guys are amazing! 

I have another question for you:

I use fetch_balance to get my assets information, assets used to contain an entryPrice field, they do not anymore! How can I get that information again? I used the 'useV2' = True and I can still get it, but if Binance deprecates this, what can I replace it with?

Thanks in advance.
Cheers 
André

**Comment by carlosmiei** [comment]
image

Hello @amdescombes actually `fetchPositions` v3 should return the `entryPrice`, if we're not returning it now might be an error parsing the response on the ccxt side, will check it ASAP

**Comment by carlosmiei** [comment]
@amdescombes Actually I just tested it and `entryPrice` is returned as expected:

```
p binanceusdm fetchPositions '["LTC/USDT:USDT"]' --sandbox
Python v3.12.3
CCXT v4.3.84
binanceusdm.fetchPositions(['LTC/USDT:USDT'])
[{'collateral': 140.91456267,
  'contractSize': 1.0,
  'contracts': 2.04,
  'datetime': '2024-07-05T07:37:18.816Z',
  'entryPrice': 69.07576601932,
  'hedged': False,
  'id': None,
  'info': {'adl': '0',
           'askNotional': '0',
           'bidNotional': '0',
           'breakEvenPrice': '74.32539390195',
           'entryPrice': '69.07576601932',
           'initialMargin': '27.15648000',
           'isolatedMargin': '0',
           'isolatedWallet': '0',
           'liquidationPrice': '0',
           'maintMargin': '0.88258560',
           'marginAsset': 'USDT',
           'markPrice': '66.56000000',
           'notional': '135.78240000',
           'openOrderInitialMargin': '0',
           'positionAmt': '2.040',
           'positionInitialMargin': '27.15648000',
           'positionSide': 'BOTH',
           'symbol': 'LTCUSDT',
           'unRealizedProfit': '-5.13216266',
           'updateTime': '1720165038816'},
  'initialMargin': 27.15648,
  'initialMarginPercentage': 0.2,
  'leverage': None,
  'liquidationPrice': None,
  'maintenanceMargin': 0.8825856,
  'maintenanceMarginPercentage': 0.0065,
  'marginMode': 'cross',
  'marginRatio': 0.0063,
  'marginType': 'cross',
  'markPrice': 66.56,
  'notional': 135.7824,
  'percentage': -18.89,
  'side': 'long',
  'stopLossPrice': None,
  'symbol': 'LTC/USDT:USDT',
  'takeProfitPrice': None,
  'timestamp': 1720165038816,
  'unrealizedPnl': -5.13216266}]
```

**Comment by amdescombes** [comment]
Thanks @carlosmiei,

Actually it is the positions in the fetch_balance call that do not return the entryPrice:
Here is my print statement:

            balances = self.eventloop.run_until_complete(self.exchange.fetch_balance())
            print(f"{json.dumps(balances['info']['positions'], indent=2, sort_keys=True)}")

and this is what I get on the new version:

  {
    "initialMargin": "23.36596934",
    "isolatedMargin": "0",
    "isolatedWallet": "0",
    "maintMargin": "0.81780888",
    "notional": "-163.56177726",
    "positionAmt": "-483",
    "positionSide": "SHORT",
    "symbol": "ADAUSDT",
    "unrealizedProfit": "-0.70458108",
    "updateTime": "1723986910746"
  },

and this is using the 'useV2' parameter:

            balances = self.eventloop.run_until_complete(self.exchange.fetch_balance(params={'useV2': True}))
            print(f"{json.dumps(balances['info']['positions'], indent=2, sort_keys=True)}")

{
  "askNotional": "0",
  "bidNotional": "0",
  "breakEvenPrice": "0.3370573419668",
  "entryPrice": "0.3371784679089",
  "initialMargin": "23.38464488",
  "isolated": false,
  "isolatedWallet": "0",
  "leverage": "7",
  "maintMargin": "0.81846253",
  "maxNotional": "1000000",
  "notional": "-163.69250604",
  "openOrderInitialMargin": "0",
  "positionAmt": "-483",
  "positionInitialMargin": "23.38464488",
  "positionSide": "SHORT",
  "symbol": "ADAUSDT",
  "unrealizedProfit": "-0.83530986",
  "updateTime": "1723986910746"
}

Thanks again !
Cheers
André

**Comment by carlosmiei** [comment]
@amdescombes I see, but `fetchBalance` uses a different endpoint from the default endpoint in `fetchPositins`, so if you use the latter, you can get the `entryPrice`

**Comment by amdescombes** [comment]
Thank you @carlosmiei,
I'll change my code to use fetchPositions
Cheers
André

**Pull request merged**